### PR TITLE
Push classification caching to be an internal detail of semantic classification. 

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/Classification/IRemoteSemanticClassificationService.cs [hunk 1]
```diff
@@ -17,7 +19,20 @@ namespace Microsoft.CodeAnalysis.Classification
...
+        /// name="textSpan"/>.  Will return an empty array not able to.
+        /// </summary>
+        /// <param name="checksum">Pass in <see cref="DocumentStateChecksums.Text"/>.  This will ensure that the cached
+        /// classifications are only returned if they match the content the file currently has.</param>
+        ValueTask<SerializableClassifiedSpans?> GetCachedSemanticClassificationsAsync(
+            DocumentKey documentKey,
+            TextSpan textSpan,
+            Checksum checksum,
+            StorageDatabase database,
+            CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Classification/IRemoteSemanticClassificationService.cs:35]: merged the semantic caching service with the semantic classification service.

File: src/Workspaces/Core/Portable/Shared/Utilities/AsyncBatchingWorkQueue`2.cs [hunk 2]
```diff
@@ -121,6 +121,23 @@ public void AddWork(IEnumerable<TItem> items)
...
+                    return;
+                }
+
+                // We're deduping items.  Only add the item if it's the first time we've seen it.
+                foreach (var item in items)
+                {
+                    if (_uniqueItems.Add(item))
+                        _nextBatch.Add(item);
+                }
+            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Utilities/AsyncBatchingWorkQueue`2.cs:139]: this was just a move for clarity.  This needs to be called under a lock.  so i moved this into the only method that calls this to make sure no one else accidentally calls this incorrectly outsdie of hte lock.

File: src/Workspaces/Remote/ServiceHub/Services/SemanticClassification/RemoteSemanticClassificationService.cs [hunk 3]
```diff
@@ -43,8 +37,22 @@ public RemoteSemanticClassificationService(in ServiceConstructionArguments argum
                 await AbstractClassificationService.AddSemanticClassificationsInCurrentProcessAsync(
                     document, span, options, temp, cancellationToken).ConfigureAwait(false);
 
-                return SerializableClassifiedSpans.Dehydrate(temp.ToImmutable());
+                if (isFullyLoaded)
+                {
+                    // WOnce full loaded, there's no need for us to keep around any of the data we cached in-memory
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/SemanticClassification/RemoteSemanticClassificationService.cs:42]:
```suggestion
                    // Once full loaded, there's no need for us to keep around any of the data we cached in-memory
```

File: src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs [hunk 4]
```diff
@@ -165,56 +169,5 @@ internal static class SemanticClassificationUtilities
                 throw ExceptionUtilities.Unreachable;
             }
         }
-
-        private static async Task AddSemanticClassificationsAsync(
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs:169]: all special logic about caching moved into the classification service itself.

File: src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs [hunk 5]
```diff
@@ -1,77 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs:1]: no more incremental analysis to cause this work to happen.  instead, we just use the pull requests on the OOP server to enqueue work to classify the full doc.

File: src/Workspaces/Core/Portable/Classification/AbstractClassificationService.cs [hunk 6]
```diff
@@ -64,6 +72,45 @@ public async Task AddSemanticClassificationsAsync(Document document, TextSpan te
...
+            // task is completed, we know that the WaitUntilFullyLoadedAsync call will have actually finished and we're
+            // fully loaded.
+            var isFullyLoadedTask = workspaceStatusService.IsFullyLoadedAsync(cancellationToken);
+            var isFullyLoaded = isFullyLoadedTask.IsCompleted && isFullyLoadedTask.GetAwaiter().GetResult();
+            return isFullyLoaded;
+        }
+
+        private static async Task<bool> TryGetCachedClassificationsAsync(
+            Document document, TextSpan textSpan, ArrayBuilder<ClassifiedSpan> result,
+            RemoteHostClient client, bool isFullyLoaded, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Classification/AbstractClassificationService.cs:91]: importantly, this work should not be pushed into the OOP client.  the idea here is that we want to make this request *without* syncing over the full solution.  so we must make the decision on the client to checked for cached results first, and if not go to normal results.

File: src/Workspaces/Core/Portable/Classification/IRemoteSemanticClassificationCacheService.cs [hunk 7]
```diff
@@ -1,35 +0,0 @@
...
-using Microsoft.CodeAnalysis.Text;
-
-namespace Microsoft.CodeAnalysis.Classification
-{
-    /// <summary>
-    /// Remote stubs used for a host to request cached semantic classifications from an OOP server.
-    /// </summary>
-    internal interface IRemoteSemanticClassificationCacheService
-    {
-        ValueTask CacheSemanticClassificationsAsync(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Classification/IRemoteSemanticClassificationCacheService.cs:19]: this method is gone completely.  the client does not choose when things are cached.

File: src/Workspaces/Core/Portable/Classification/IRemoteSemanticClassificationCacheService.cs [hunk 8]
```diff
@@ -1,35 +0,0 @@
...
-        /// name="textSpan"/>.  Will return an empty array not able to.
-        /// </summary>
-        /// <param name="checksum">Pass in <see cref="DocumentStateChecksums.Text"/>.  This will ensure that the cached
-        /// classifications are only returned if they match the content the file currently has.</param>
-        ValueTask<SerializableClassifiedSpans?> GetCachedSemanticClassificationsAsync(
-            DocumentKey documentKey,
-            TextSpan textSpan,
-            Checksum checksum,
-            StorageDatabase database,
-            CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Classification/IRemoteSemanticClassificationCacheService.cs:33]: this method moved to the normal classification service.

File: src/Workspaces/Remote/ServiceHub/Services/SemanticClassification/RemoteSemanticClassificationService.cs [hunk 9]
```diff
@@ -43,6 +37,17 @@ public RemoteSemanticClassificationService(in ServiceConstructionArguments argum
                 await AbstractClassificationService.AddSemanticClassificationsInCurrentProcessAsync(
                     document, span, options, temp, cancellationToken).ConfigureAwait(false);
 
+                if (isFullyLoaded)
+                {
+                    // Once full loaded, there's no need for us to keep around any of the data we cached in-memory
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/SemanticClassification/RemoteSemanticClassificationService.cs:42]:
```suggestion
                    // Once fully loaded, there's no need for us to keep around any of the data we cached in-memory
```

**Review by dibarbet — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@allisonchou this is available for you now. Semantic classification caching should now work transparently.

**Comment by allisonchou** [comment]
Awesome, thanks so much Cyrus!

**Pull request merged**